code/messaging/src/main/java/com/adobe/marketing/mobile/messaging/internal/EdgePersonalizationResponseHandler.java

```diff
@@ -163,128 +205,325 @@ void fetchMessages(final List<Surface> surfaces) {
+        final List<LaunchRule> inAppRules = parsedRules.get(InboundType.INAPP);
+        final List<LaunchRule> feedRules = parsedRules.get(InboundType.FEED);
+
+        // handle in-app message rules
+        if (!MessagingUtils.isNullOrEmpty(inAppRules)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "The personalization:decisions response contains InApp message definitions.");
+        }
+        if (clearExistingRules) {
+            Log.debug(LOG_TAG, SELF_TAG, "Loading %d rule(s) into the rules engine.", inAppRules != null ? inAppRules.size() : 0);
+            launchRulesEngine.replaceRules(inAppRules);
+        } else {
+            Log.debug(LOG_TAG, SELF_TAG, "Added %d rule(s) into the rules engine.", inAppRules != null ? inAppRules.size() : 0);
+            launchRulesEngine.addRules(inAppRules);
+        }
+
+        // handle feed rules
+        if (!MessagingUtils.isNullOrEmpty(feedRules)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "The personalization:decisions response contains feed message definitions.");
+        }
+        if (clearExistingRules) {
+            Log.debug(LOG_TAG, SELF_TAG, "Loading %d rule(s) into the feed rules engine.", feedRules != null ? feedRules.size() : 0);
+            feedRulesEngine.replaceRules(feedRules);
+        } else {
+            Log.debug(LOG_TAG, SELF_TAG, "Added %d rule(s) into the feed rules engine.", feedRules != null ? feedRules.size() : 0);
+            feedRulesEngine.addRules(feedRules);
         }
 
-        final String appSurface = getAppSurface();
-        Log.trace(LOG_TAG, SELF_TAG, "Loading in-app message definitions from personalization:decisions network response.");
-        processPropositions(propositions, clearExistingRules, true, appSurface);
+        inboundMessages = feedRulesEngine.evaluate(edgeResponseEvent);
+        if (!MapUtils.isNullOrEmpty(inboundMessages)) {
+            updateInboundMessages(inboundMessages, requestedSurfaces);
+        }
+
+        final Map<Surface, List<Proposition>> requestedPropositions = retrievePropositions(requestedSurfaces);
+        if (MapUtils.isNullOrEmpty(requestedPropositions)) {
+            Log.trace(MessagingConstants.LOG_TAG, SELF_TAG, "Not dispatching a notification event, personalization:decisions response does not contain propositions.");
+            return;
+        }
+
+        // dispatch an event with the propositions received from the remote
+        final Map<String, Object> eventData = new HashMap<>();
+        final List<Map<String, Object>> convertedPropositions = new ArrayList<>();
+        for (final Map.Entry<Surface, List<Proposition>> propositionEntry : requestedPropositions.entrySet()) {
+            for (final Proposition proposition : propositionEntry.getValue()) {
+                convertedPropositions.add(proposition.toEventData());
+            }
+        }
+        eventData.put(PROPOSITIONS, convertedPropositions);
```

Comment:
Since you updated how propositions are passed in the event data, please ensure it is being extracted/ read correctly in the public APIs.

Reply:
i'll have this updated in the feed demo pr